### PR TITLE
Remove awesome nested set gem and dependencies [OFN-11636]

### LINKED ISSUE #6033: Errors when deleting (and creating?) taxonomies
## Description

While testing #5868 two BugsnagJS errors were triggered:

i) when creating a new taxon/child

https://app.bugsnag.com/yaycode/open-food-network-js-uk/errors/5f5f9c8d35c98900171659b8?event_id=5f5f9c8d005dfcffa4760000&i=sk&m=nw

Nothing visible to the user in this case, only BugsnagJS triggered. Could not reproduce this error.

ii.a) when attempting to delete a child-taxonomy which is is attributed to a product.

https://app.bugsnag.com/yaycode/open-food-network-js-uk/errors/5f5fa317f26a69001738bcf0?event_id=5f5fa317005e00fe89a20000&i=sk&m=nw

An the page freezes (endless spinner), error 422 is seen on the console (see pic below). This is reproducible, see steps to reproduce.

ii.b) when attempting to delete a main-taxonomy which is is attributed to a product.

https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/5f60cf8b37940e00177318ae?event_id=5f60cf8b005e05d971660000&i=sk&m=nw

## Expected Behavior

i) Creating taxonomies should not trigger errors on BugsnagJS

ii) It should not be possible to delete taxonomies when these are attributed to products, right? This is a mandatory field for products.
So, perhaps an improvement would be an informative message stating that the taxonomy X is attributed to product Y and cannot be deleted?

## Actual Behaviour

i) BugnsagJS was triggered, but nothing unusual is shown to the user. 

ii) Cannot delete a taxonomy (main or child) attributed to a product. Attempting to do so:
- for a child-taxonomy: screen freezes with error 422 (console) and BugsnagJS error - see ii.a) above. Refreshing the page shows the previously available taxonomies
- for a main taxonomy: snail/error 500 and Bugsnag error - see ii.b) above

## Steps to Reproduce

i) could not reproduce this error 

ii)

1. Log-in as superadmin. 
2. Go to /admin/taxonomies
3. Create a taxonomy and child taxonomies
4. Create new products with these taxonomies or add them to existing products.
5. Go back to /admin/taxonomies
6. Try to delete a child taxonomy - see error ii.a)
7. Try to delete a main taxonomy - see error ii.b)

## Animated Gif/Screenshot

Error ii.a)

image

Error ii.b)

image

## Workaround

- Delete the product which has the taxonomy you want to delete. This is currently not always possible, due to #5947 
- If you can't delete the product then temporarily assign another taxonomy to that product.
Deleting should work after one of these steps. 

## Severity

bug-s4

## Your Environment

* Version used: v3.2.6
* Browser name and version: Firefox 80
* Operating System and version (desktop or mobile): Desktop/Ubuntu 20.04

## Possible Fix

Related to #6030, 5252 and #5947.

### LINKED ISSUE #9385: Super admin: Renaming the header taxon not working when right clicking and selecting 'edit'
## Description

Follow up of #9255.

Renaming the header taxon is not working when right clicking and selecting 'edit'. The problem is that the input field at the top of the page is not updated accordingly when updating the 'edit' page.

Also related:
- #9386  
- #9387 

## Expected Behavior

The changes should take affect as soon as 'update' has been clicked on the 'edit' page of the header taxon.

## Actual Behaviour

The changes are not applied. The old name persists.

## Steps to Reproduce

1. As super admin user go to admin/taxonomies.
2. Select the 'edit' button on the right of a taxon.
3. Right click on the header taxon and select 'edit'.
4. Change the name of the taxon.
5. Click update.
6. Click update on the next page as well.
7. See that the change didn't take effect.

## Animated Gif/Screenshot

This screen capture demonstrates all three option to rename the header taxon:
- using the input field above the taxons --> OK
- using right click rename --> OK
- using right click edit --> **not ok**

Peek 2022-06-17 15-03

## Workaround

Use one of the working methods.

## Severity
bug-s5: we can live with it, only a few users impacted

## Your Environment

* Version used: staged [68bd1df312014a0ffaef99d24a3d1ff2751cb546](https://github.com/openfoodfoundation/openfoodnetwork/pull/9255/commits/68bd1df312014a0ffaef99d24a3d1ff2751cb546) 
* Browser name and version: Firefox 101.0.1 (64-Bit)
* Operating System and version (desktop or mobile): Desktop Ubuntu 20.04 LTS

### LINKED ISSUE #9386: Super admin: Changing the order of taxons is not working
## Description

Follow up of #9255.

Changing the order of taxons is not working. The order is currently not being used anywhere in the front end, but it may be used at a later point (see [this comment](https://github.com/openfoodfoundation/openfoodnetwork/pull/9255#issuecomment-1158906907)).

Also related:
- #9385 
- #9387 

## Expected Behavior

Changing the order of taxons and clicking 'update' should keep the new order of the taxons.

## Actual Behaviour

The order is not kept. The old order is displayed again.

## Steps to Reproduce

1. As super admin go to admin/taxonomies.
2. Select the edit button at the right of a taxon.
3. Change the order of sub taxons using drag and drop.
4. Click update.
5. Select the edit button at the right of the taxon again.
6. See that the old order is being restored.

## Animated Gif/Screenshot

Peek 2022-06-17 15-13

## Workaround

None.

## Severity
bug-s5: we can live with it, only a few users impacted

## Your Environment

* Version used: staged [68bd1df312014a0ffaef99d24a3d1ff2751cb546](https://github.com/openfoodfoundation/openfoodnetwork/pull/9255/commits/68bd1df312014a0ffaef99d24a3d1ff2751cb546) 
* Browser name and version: Firefox 101.0.1 (64-Bit)
* Operating System and version (desktop or mobile): Desktop Ubuntu 20.04 LTS

### LINKED ISSUE #9387: Super admin: Deleting the last taxon or header taxon using right click 'remove' does not work
## Description

Follow up of #9255.

Deleting the last taxon or header taxon using right click 'remove' does not work. After clicking 'update' and checking again, the deleted taxons are back.

Also related:
- #9385 
- #9386 

## Expected Behavior

Deleting taxons using right click 'remove' should remove the taxon.

## Actual Behaviour

After clicking 'update' and checking again, the deleted taxons are back.

## Steps to Reproduce

**Header taxons:**
1. As super admin go to admin/taxonomies.
2. Select the 'edit' button on the right of a taxon.
3. Right click on the header taxon and select 'remove'.
4. Confirm the dialog 'Are you sure?'.
5. Observe that there is a never ending 'Loading...' sign at the top of the page.
6. Click update.
7. Observe that the header taxon is still in the list.

Peek 2022-06-17 15-16

**Last taxon:**
1. As super admin go to admin/taxonomies.
2. Select the 'edit' button on the right of a taxon.
3. Right click on one of the sub taxons and select 'remove'.
4. Confirm the dialog 'Are you sure?'.
5. Repeat 3. and 4. until there is only one sub taxon left.
6. Repeat 3. and 4. for the last sub taxon.
7. Observe that there is a never ending 'Loading...' sign at the top of the page.
9. Click update.
10. Select the 'edit' button on the right of the taxon again.
11. Observe that the last taxon is still in the list.

Peek 2022-06-17 15-18

## Animated Gif/Screenshot

See above.

## Workaround

None.

## Severity
bug-s5: we can live with it, only a few users impacted

## Your Environment

* Version used: staged [68bd1df312014a0ffaef99d24a3d1ff2751cb546](https://github.com/openfoodfoundation/openfoodnetwork/pull/9255/commits/68bd1df312014a0ffaef99d24a3d1ff2751cb546) 
* Browser name and version: Firefox 101.0.1 (64-Bit)
* Operating System and version (desktop or mobile): Desktop Ubuntu 20.04 LTS

### LINKED ISSUE #9677: Remove `jquery.jstree` from the taxonomy page
## Description
We should remove `jquery.jstree` from the `/admin/taxonomies/[TAXON_ID]/edit` page:
 - we should use a package that is imported via package.json and yarn
 - this trigger some errors for images not found (`GET http://localhost:3000/throbber.gif 404 (Not Found)`) 

#### Context
https://github.com/openfoodfoundation/openfoodnetwork/pull/9675/commits/cf8bab97e8944b18d01efaf8f7a755d9fac3ab02

### LINKED ISSUE #11636: Remove gem awesome_nested_set
### What we should change and why (this is tech debt)

It's providing a tree structure for Spree::Taxon but we don't use the tree structure. I think that the UI is actually not saving it when you try. If we ever wanted a tree structure in the future, it would be better to use [ancestry](https://github.com/stefankroes/ancestry) for it's simpler design and better update performance.

There are several related issues that may be resolved at the same time:

* #9677
* #9386
* #6033

### Context

* #11631

### Impact and timeline

No significant impact at the moment. But simplifying code always reduces maintenance and performance.

### PR DESCRIPTION
**Pull request by wandji20**
#### What? Why?

- Closes #11636
- Closes #6033
- Closes #9385
- Closes #9386
- Closes #9387
- Closes #9677

Awesomenested set and related dependencies where removed
- Taxonomy model (deleted)
- Attributes: `parent_id`, `lft`, and  `rgt` where removed from Taxon model
- Taxonomy controller (including API/V0) and specs deleted
- Taxon controller ((including API/V0) and specs updated to reflect model change and all CRUD actions added
- Taxons replaced Taxonomies in admin setting page

#### What should we test?
- Admin setting page should work properly with new taxons options
- Taxnonomies should be removed from admin settings page
- All crud actions should be available in adminnn settings for taxons
#### Video
[Screencast from 07-08-2024 15:48:29.webm](https://github.com/user-attachments/assets/2f2b6e04-564b-45bc-bfc8-d0dc7cbccf32)

- Visit ... page.
http://localhost:3000/admin/general_settings/edit

#### Release notes

- Remove unused tree structure on product categories

Changelog Category (reviewers may add a label for the release notes):

- [x] User facing changes
- [x] API changes (V0)
- [ ] Technical changes only

Remove unused tree structure on product categories

#### Dependencies
N/A

#### Documentation updates
N/A

### CODE REVIEW

File: db/migrate/20240806135838_drop_spree_taxonomies_table.rb [hunk 1]
```diff
@@ -0,0 +1,14 @@
+class DropSpreeTaxonomiesTable < ActiveRecord::Migration[7.0]
+  def change
+    # Remove columns
+    remove_column :spree_taxons, :lft
+    remove_column :spree_taxons, :rgt
+
+    # Remove references
+    remove_reference :spree_taxons, :parent, index: true, foriegn_key: true
+    remove_reference :spree_taxons, :taxonomy, index: true, foriegn_key: true
```
[the remark this answers is not in the document]
Comment (mkllnk) [reply db/migrate/20240806135838_drop_spree_taxonomies_table.rb:9]: I'll add a commit to make the migration reversible.

**Review by mkllnk — APPROVED** [review]
Excellent work! Thank you.

It would have been easier to review if you had created several smaller commits to guide us through the changes step by step.

I'm adding one commit to fix up the migration but otherwise it's all good. There's one leftover jquery plugin now that we can remove as well. Would you like to tackle that?

**Comment by mkllnk** [comment]
Oh, and it looks like you need to rebase this pull request on the latest master to resolve conflicts.

**Comment by mkllnk** [comment]
While testing I noticed that the permalink is not visible. Setting a new value doesn't work either. I don't actually think that it's used anywhere. Can you remove at least the input field from the form?

File: app/controllers/spree/admin/taxons_controller.rb [hunk 2]
```diff
@@ -2,122 +2,70 @@
...
-        respond_with(@taxon) do |format|
-          format.html { redirect_to spree.edit_admin_taxonomy_url(@taxonomy) }
-          format.json { render json: @taxon.to_json }
+        respond_to do |format|
+          format.html {
+            flash[:success] = flash_message if status == :ok
+            flash[:error] = flash_message if status == :unprocessable_entity
+            redirect_to admin_taxons_path
+          }
+          format.turbo_stream {
```
Comment (rioug) [inline app/controllers/spree/admin/taxons_controller.rb:52]: are we using turbo with this ?

**Review by rioug — APPROVED** [review]
Awesome, thanks ! :pray: 
I'll second Maikel here, it would have been easier to follow if it was split in multiple small commit.
I think we'll need to flag this to the instance manager when it get released, they might want to review their taxonomy afterward.

**Comment by drummer83** [comment]
Hi @wandji20,
Thank you for yet another PR! :muscle: 
I have tested it on our servers.

### Configuration tabs in the instance settings
Before:
image

After:
image

### Taxonomies vs. Product Categories page
Before:
image

image

image

After:
image

### Taxonomy edit vs. Product Category edit page
Before:
image

After:
image

### Header taxonomies and sub taxons listed as Product Categories
After:
image

## Additional test cases
- New page 'Product Categories'. :heavy_check_mark: 
- Old page 'Taxonomies' is gone. :heavy_check_mark: 
- All previous header taxonomies are listed as product categories now. :heavy_check_mark: 
- All previous taxons and sub taxons are listed as product categories. :heavy_check_mark: 
- Create new product category. :heavy_check_mark: 
- Edit product category. :heavy_check_mark: 
- Can't delete a product category with assigned products. :heavy_check_mark: 
image
- Can delete a product category without assigned products. :heavy_check_mark: 
- Product categories are listed in an alphabetical order. :heavy_check_mark: 
- Permalink has been removed. :heavy_check_mark: 

### Open (nice to have in a follow-up)
- Improve error message (`spree.admin.taxons.destroy.delete_taxon.error`): `Unable to delete the product category due to assigned products.` :question: 
- Is the key `spree_classification_primary_taxon_error` still used anywhere? :question: 
- Is the key `spree.new_taxon` still used anywhere? :question: 
- Is the key `spree.api.invalid_taxonomy_id` still used anywhere? :question: 

## Task
:exclamation: Inform instance managers about this change.

## Conclusion
Everything I tested is working well. Great job!
I only have minor questions, which can be addressed in a follow up, if at all.

This PR is ready to go. Merging! :rocket: 
Thanks! :tada:

**Comment by drummer83** [comment]
Created #12773 for follow up issues.

**Pull request merged**